### PR TITLE
Osiris/ext ethapi

### PR DESCRIPTION
**Pull request by 0xOsiris**
Addresses #72

This PR creates an RPC server in the builder that Extends the EthApi with `eth_sendRawTransactionConditional`. This RPC method validates the `ConditionalOptions` at the RPC layer prior to inserting the transaction into the transaction pool. 

I have excluded Transaction forwarding to the `sequencer_http` from the implementation as this will likely be used by Pbh bundles, but open to any feedback. 

Integration CI Test: https://github.com/worldcoin/world-chain/actions/runs/12244869012

### CODE REVIEW

File: world-chain-builder/crates/assertor/src/main.rs [hunk 1]
```diff
@@ -128,11 +128,28 @@ where
...
+            .await?;
+        queue.push(PendingTransactionBuilder::new(
+            builder_provider.root(),
+            tx_hash,
+        ));
+    }
+
     let futs = queue.into_iter().map(|builder| async {
         let hash = builder.tx_hash().clone();
         let receipt = builder.get_receipt().await;
```
Comment (0xForerunner) [inline world-chain-builder/crates/assertor/src/main.rs]:
```suggestion
    let num_transactions = fixture.fixture.len();
    let half = num_transactions / 2;

    let builder_provider_clone = builder_provider.clone();

    stream::iter(fixture.fixture.iter().enumerate())
        .map(Ok)
        .try_for_each_concurrent(CONCURRENCY_LIMIT, move |(index, transaction)| {
            let builder_provider = builder_provider_clone.clone();
            async move {
                let tx = if index < half {
                    // First half, use send_raw_transaction
                    builder_provider.send_raw_transaction(transaction).await?
                } else {
                    // Second half, use eth_sendRawTransactionConditional
                    let rlp_hex = hex::encode_prefixed(transaction);
                    let tx_hash = builder_provider
                        .client()
                        .request(
                            "eth_sendRawTransactionConditional",
                            (rlp_hex, ConditionalOptions::default()),
                        )
                        .await?;
                    PendingTransactionBuilder::new(builder_provider.root(), tx_hash)
                };
                let hash = *tx.tx_hash();
                let receipt = tx.get_receipt().await;
                assert!(receipt.is_ok());
                debug!(
                    receipt = ?receipt.unwrap(),
                    hash = ?hash,
                    index = index,
                    "Transaction Receipt Received"
                );
                Ok::<(), eyre::Report>(())
            }
        })
        .await?;
```

**Review by 0xForerunner — APPROVED** [review]
Looks good man!

File: world-chain-builder/src/rpc/bundle.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        if let Some(min_timestamp) = options.timestamp_min {
+            if min_timestamp > latest.timestamp {
+                return Err(ErrorCode::from(-32003).into());
+            }
+        }
+
+        if let Some(max_timestamp) = options.timestamp_max {
+            if max_timestamp <= latest.timestamp {
```
Comment (0xKitsune) [inline world-chain-builder/src/rpc/bundle.rs:113]: Should this be `if max_timestamp < latest.timestamp`? If the timestamp is equal to the max timestamp I am under the impression that the bundle is still valid?

File: world-chain-builder/src/rpc/bundle.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        self.validate_known_accounts(options.known_accounts, latest.header.number.into())?;
+
+        if let Some(min_block) = options.block_number_min {
+            if min_block > latest.number {
+                return Err(ErrorCode::from(-32003).into());
+            }
+        }
+
+        if let Some(max_block) = options.block_number_max {
+            if max_block <= latest.number {
```
Comment (0xKitsune) [inline world-chain-builder/src/rpc/bundle.rs:101]: Should this be `if max_block < latest.number`?

File: world-chain-builder/src/rpc/bundle.rs (hunk 2 above)
Comment (0xOsiris) [reply world-chain-builder/src/rpc/bundle.rs:113]: If `max_timestamp == latest.timestamp` then the next block will always have the property `max_timestamp < next_block.timestamp` which would be invalid. Was the thought

Comment (0xKitsune) [reply world-chain-builder/src/rpc/bundle.rs:113]: Sgtm makes sense! Consider adding a comment here as well.

Comment (0xOsiris) [reply world-chain-builder/src/rpc/bundle.rs:113]: I believe we are picking up the last confirmed block here, not the latest pending. But let me double check.

> Yeah this is the latest block, not the latest pending block.

File: world-chain-builder/src/rpc/bundle.rs (hunk 3 above)
Comment (0xOsiris) [reply world-chain-builder/src/rpc/bundle.rs:101]: I think it should be `<=` as the pending block will always be > then `latest.number`

File: world-chain-builder/src/rpc/bundle.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+        for (address, storage) in known_accounts.into_iter() {
+            match storage {
+                AccountStorage::Slots(slots) => {
+                    for (slot, value) in slots.into_iter() {
+                        let current = state.storage(address, slot.into()).map_err(|e| {
+                            ErrorObject::owned(
+                                ErrorCode::InternalError.code(),
+                                e.to_string(),
+                                Some(""),
```
Comment (0xKitsune) [inline world-chain-builder/src/rpc/bundle.rs:141]: In cases where we are returning an empty string here, can we just return `None` instead?

**Review by 0xKitsune — APPROVED** [review]

**Pull request merged**